common/lib/xmodule/xmodule/partitions/partitions_service.py

```diff
@@ -3,41 +3,99 @@
+
+
+def _create_enrollment_track_partition(course):
+    """
+    Create and return the dynamic enrollment track user partition.
+    If it cannot be created, None is returned.
+    """
+    try:
+        enrollment_track_scheme = UserPartition.get_scheme("enrollment_track")
+    except UserPartitionError:
+        log.warning("No 'enrollment_track' scheme registered, EnrollmentTrackUserPartition will not be created.")
+        return None
+
+    used_ids = set(p.id for p in course.user_partitions)
+    if ENROLLMENT_TRACK_PARTITION_ID in used_ids:
+        log.warning("Cannot add 'enrollment_track' partition, as ID {id} is assigned to {partition}".format(
+            id=ENROLLMENT_TRACK_PARTITION_ID,
+            partition=_get_partition_from_id(course.user_partitions, ENROLLMENT_TRACK_PARTITION_ID).name)
+        )
+        return None
+
+    partition = enrollment_track_scheme.create_user_partition(
+        id=ENROLLMENT_TRACK_PARTITION_ID,
+        name=_(u"Enrollment Track Partition"),
+        description=_(u"Partition for segmenting users by enrollment track"),
+        parameters={"course_id": unicode(course.id)}
+    )
+    return partition
 
 
 class PartitionService(object):
     """
-    This is an XBlock service that assigns tracks which groups users are in for various
-    user partitions.  It uses the provided user_tags service object to
-    persist the assignments.
+    This is an XBlock service that returns information about the user partitions associated
+    with a given course.
     """
-    __metaclass__ = ABCMeta
 
-    @abstractproperty
+    def __init__(self, course_id, track_function=None, cache=None):
+        self._course_id = course_id
+        self._track_function = track_function
+        self._cache = cache
+
+    def get_course(self):
+        """
+        Return the course instance associated with this PartitionService.
+        Returns:
+
+        """
+        return modulestore().get_course(self._course_id)
+
+    @property
     def course_partitions(self):
         """
         Return the set of partitions assigned to self._course_id
         """
-        raise NotImplementedError('Subclasses must implement course_partition')
-
-    def __init__(self, user, course_id, track_function=None, cache=None):
-        self._user = user
-        self._course_id = course_id
-        self._track_function = track_function
-        self._cache = cache
+        return get_course_user_partitions(self.get_course())
 
-    def get_user_group_id_for_partition(self, user_partition_id):
+    def get_user_group_id_for_partition(self, user, user_partition_id):
         """
         If the user is already assigned to a group in user_partition_id, return the
         group_id.
 
         If not, assign them to one of the groups, persist that decision, and
         return the group_id.
 
-        If the group they are assigned to doesn't exist anymore, re-assign to one of
```

Comment:
Note to reviewers: I don't see that the code actually does this. Perhaps the random user partition scheme does, but that is implementation-specific to a specific scheme.